wp-content/themes/core/components/tabs/Tabs_Controller.php

```diff
@@ -0,0 +1,244 @@
+<?php
+
+namespace Tribe\Project\Templates\Components\tabs;
+
+use Tribe\Libs\Utils\Markup_Utils;
+use Tribe\Project\Templates\Components\Abstract_Controller;
+use Tribe\Project\Templates\Models\Tab as Tab_Model;
+
+/**
+ * Class Controller
+ *
+ * @package Tribe\Project\Templates\Components\tabs
+ */
+class Tabs_Controller extends Abstract_Controller {
+	public const TABS               = 'tabs';
+	public const CLASSES            = 'classes';
+	public const ATTRS              = 'attrs';
+	public const LAYOUT             = 'layout';
+	public const TOGGLE_CLASSES     = 'toggle_classes';
+	public const TAB_BUTTON_CLASSES = 'tab_button_classes';
+	public const TAB_PANEL_CLASSES  = 'tab_panel_classes';
+	public const LAYOUT_HORIZONTAL  = 'horizontal';
+	public const LAYOUT_VERTICAL    = 'vertical';
+
+	/**
+	 * @var Tab_Model[] The collection of tabs to render. Each item should be a \Tribe\Project\Templates\Models\Tab object.
+	 */
+	private array $tabs;
+
+	private array $classes;
+	private array $attrs;
+	private string $layout;
+	private array $toggle_classes;
+	private array $tab_button_classes;
+	private array $tab_panel_classes;
+	private string $tablist_id;
+	private array $tab_buttons = [];
+	private array $tab_panels = [];
+
+	/**
+	 * Controller constructor.
+	 *
+	 * @param array $args
+	 */
+	public function __construct( array $args = [] ) {
+		$args = $this->parse_args( $args );
+
+		$this->tabs               = (array) $args[ self::TABS ];
+		$this->classes            = (array) $args[ self::CLASSES ];
+		$this->attrs              = (array) $args[ self::ATTRS ];
+		$this->layout             = (string) $args[ self::LAYOUT ];
+		$this->toggle_classes     = (array) $args[ self::TOGGLE_CLASSES ];
+		$this->tab_button_classes = (array) $args[ self::TAB_BUTTON_CLASSES ];
+		$this->tab_panel_classes  = (array) $args[ self::TAB_PANEL_CLASSES ];
+		$this->tablist_id         = uniqid( 'c-tabs__tablist--' );
+
+		$this->init_tabs();
+	}
+
+	/**
+	 * @return array
+	 */
+	protected function defaults(): array {
+		return [
+			self::TABS               => [],
+			self::CLASSES            => [],
+			self::ATTRS              => [],
+			self::LAYOUT             => self::LAYOUT_HORIZONTAL,
+			self::TOGGLE_CLASSES     => [],
+			self::TAB_BUTTON_CLASSES => [],
+			self::TAB_PANEL_CLASSES  => [ 's-sink', 't-sink' ],
+		];
+	}
+
+	/**
+	 * @return array
+	 */
+	protected function required(): array {
+		return [
+			self::CLASSES            => [ 'c-tabs' ],
+			self::ATTRS              => [ 'data-js' => 'c-tabs' ],
+			self::TOGGLE_CLASSES     => [ 'c-tabs__tablist-toggle' ],
+			self::TAB_BUTTON_CLASSES => [ 'c-tabs__tab' ],
+			self::TAB_PANEL_CLASSES  => [ 'c-tabs__tabpanel' ],
+		];
+	}
+
+	/**
+	 * Loop through the tabs provided setup the tab button and tab panel components arguments for each.
+	 */
+	private function init_tabs() {
+		foreach ( $this->tabs as $index => $tab ) {
```

Comment:
Interested in your thoughts on how I approached accordion. I setup the header_id and content_ids in the constructor for the Accordion Model. https://github.com/moderntribe/square-one/blob/epic/fidgety-feet/wp-content/plugins/core/src/Blocks/Types/Accordion/Accordion_Model.php#L36-L37 Wondering if we don't want to adopt something similar to what you have done here if there is anything specific that needs added other than a unique id.

Reply:
Yep. I like this arch a bit better because it focuses all the setup for the repeater things together in once place rather than passing generated uuids along with the "real" data into the component.